core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -999,19 +1003,19 @@ private void assertCanWrite() throws IOException {
         }
     }
 
+    // package private for testing
+    static final String TEMP_FILE_NAME = ".es_temp_file";
+
     private static void tryWriteTempFile(Path path) throws IOException {
         if (Files.exists(path)) {
-            Path resolve = path.resolve(".es_temp_file");
-            boolean tempFileCreated = false;
+            Path resolve = path.resolve(TEMP_FILE_NAME);
             try {
+                // delete any lingering file from a previous failure
+                Files.deleteIfExists(resolve);
```

Comment:
maybe move all of the non-write operations out of the try block then the exception makes sense again